text/0000-run-to-completion-async-fn.md

```diff
@@ -0,0 +1,602 @@
+will lead to `panic`s or `abort`s, since there exists not way to handle a
+cancellation in a reasonable fashion. By introducing run-to-completion function
+we will at least gain a feature on type-system level that will help to disambiguate
+functions which can safely be synchronously cancelled and ones which need to
+run to completion.
+
+### The new trait is `unsafe` by default
+
+The new proposed trait requires the implementation of an `unsafe` method. The
+`unsafe` annotation here is mainly required to enforce a special contract with
+the caller which is currently not achievable within Rusts safe type system.
+
+This is unfortunate, since we strive to minimize the use of `unsafe` code within
+Rust codebases. The `unsafe` annotation also allows implementors of the trait
+to use any arbitrary unsafe code, incl. code which could potentially lead to
+memory unsafety issues. This is not necessarily intended, because we want
+`unsafe` here mainly in order to enforce a contract with the caller.
+
+However we do not expect a lot of people to implement `RunToCompletionFuture`
+themselves. It should mainly be generated by the compiler. In addition this low
+level primitives like IO completion based socket types would need to be implemented
+by runtime authors - but those will need unsafe code anyway (they need to pass
+raw pointers to underlying C code or the kernel for an amount of time which can
+not be checked by lifetimes). Most users are purely expected to use high level
+`async completion fn`s, which will not require unsafe code. 
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The proposal presented here provides a way to add run to completion semantics
+for asynchronous functions in a backwards compatible and minimally invasive way.
+Since neither the way an async function desugars into a generator/state-machine nor
+the way wakeups (the task system) are performed are touched, the effort to integrate
+this mechanism should be manageable.
+
+The mechanism will allow us to provide true zero cost abstractions for completion
+based operations. Those already exist in other languages (C++, Zig, C#, Kotlin,
+etc), but not yet in Rust.
+
+`async completion fn` will also introduce an async function type which behaves
+more simliar to normal functions that current `async fn`s do. It might therefore
+have been interesting to see `async completion fn` as the new "default" async
+function type. However this is not possible anymore, since `async fn` is already
+stabilized.
+
+And while the introduction of another type of async function type sounds like
+as an additional complexity at first, it also has its benefits:
+
+Users will continue to be able to write simple and safe `Future` implementations
+with cancel-anytime semantics manually, and be able to use them in combination
+with powerful flow-control macros like `select!` and `join!`. A lot of types
```

Comment:
The big issue I see is that this affects leaf futures and therefore would require the entire future chain to be a completion future. Which is quite unfortunate.

Reply:
I guess if we provided some `tokio::spawn_completion` type then you could use its join handle to "forget" about that. That should be a somewhat and minimal way to not have to funnel completion up the chain.

Reply:
You mean and leave it running in the background?
I think in order to avoid having those runaway tasks we will implement the `scope` proposal to work based on this.

Reply:
Basically, what I have now for uring does this, even if the future is canceled it will still run in the background. But I think that is useful to mix with other control flow.